packages/protocol/contracts/common/FixidityLib.sol

```diff
@@ -0,0 +1,338 @@
+   * @dev Test newFixed(0) returns 0
+   * Test newFixed(1) returns fixed1()
+   * Test newFixed(maxNewFixed()) returns maxNewFixed() * fixed1()
+   * Test newFixed(maxNewFixed()+1) fails
+   */
+  function newFixed(uint256 x)
+    internal
+    pure
+    returns (Fraction memory)
+  {
+    require(x <= maxNewFixed());
+    return Fraction(x * FIXED1_UINT);
+  }
+
+  /**
+   * @notice Converts a uint256 in the fixed point representation of this
+   * library to a non decimal. All decimal digits will be truncated.
+   */
+  function fromFixed(Fraction memory x)
+    internal
+    pure
+    returns (uint256)
+  {
+    return x.value / FIXED1_UINT;
+  }
+
+  /**
+   * @notice Converts two uint256 representing a fraction to fixed point units,
+   * equivalent to multiplying dividend and divisor by 10^digits().
+   * @dev
+   * Test newFixedFraction(maxFixedDividend()+1,1) fails
+   * Test newFixedFraction(1,maxFixedDividend()+1) fails
+   * Test newFixedFraction(1,0) fails
+   * Test newFixedFraction(0,1) returns 0
+   * Test newFixedFraction(1,1) returns fixed1()
+   * Test newFixedFraction(maxFixedDividend(),1) returns maxFixedDividend()*fixed1()
+   * Test newFixedFraction(1,fixed1()) returns 1
+   */
+  function newFixedFraction(
+    uint256 numerator,
+    uint256 denominator
+  )
+    internal
+    pure
+    returns (Fraction memory)
+  {
+    require(numerator <= maxNewFixed());
+    require(denominator <= maxNewFixed());
+    require(denominator != 0);
+    Fraction memory convertedNumerator = newFixed(numerator);
+    Fraction memory convertedDenominator = newFixed(denominator);
+    return divide(convertedNumerator, convertedDenominator);
+  }
+
+  /**
+   * @notice Returns the integer part of a fixed point number.
+   * @dev
+   * Test integer(0) returns 0
+   * Test integer(fixed1()) returns fixed1()
+   * Test integer(newFixed(maxNewFixed())) returns maxNewFixed()*fixed1()
+   */
+  function integer(Fraction memory x) internal pure returns (Fraction memory) {
+    return Fraction((x.value / FIXED1_UINT) * FIXED1_UINT); // Can't overflow
+  }
+
+  /**
+   * @notice Returns the fractional part of a fixed point number.
+   * In the case of a negative number the fractional is also negative.
+   * @dev
+   * Test fractional(0) returns 0
+   * Test fractional(fixed1()) returns 0
+   * Test fractional(fixed1()-1) returns 10^24-1
+   */
+  function fractional(Fraction memory x) internal pure returns (Fraction memory) {
+    return Fraction(x.value - (x.value / FIXED1_UINT) * FIXED1_UINT); // Can't overflow
+  }
+
+  /**
+   * @notice x+y. If any operator is higher than maxFixedAdd() it
+   * might overflow.
+   * @dev
+   * Test add(maxFixedAdd(),maxFixedAdd()) returns maxUint256()-1
+   * Test add(maxFixedAdd()+1,maxFixedAdd()+1) fails
+   * Test add(maxUint256(),maxUint256()) fails
+   */
+  function add(Fraction memory x, Fraction memory y) internal pure returns (Fraction memory) {
+    uint256 z = x.value + y.value;
```

Comment:
same code as `SafeMath`, does it makes sense to use it?
```
    function add(uint256 a, uint256 b) internal pure returns (uint256) {
        uint256 c = a + b;
        require(c >= a, "SafeMath: addition overflow");

        return c;
    }
```

Also. I think it would be nice to have error messages on every `require`